src/twister2/device/native_simulator_adapter.py

```diff
@@ -10,91 +17,124 @@
-            t1.start()
-            t2.start()
-            t1.join()
-            t2.join()
-        except subprocess.CalledProcessError as e:
-            logger.exception('Flashing failed due to error %s', e)
-            raise
+        except subprocess.SubprocessError as e:
+            logger.error('Running simulation failed due to subprocess error %s', e)
+            self._exc = TwisterRunException(e.args)
+        except FileNotFoundError as e:
+            logger.error(f'Running simulation failed due to file not found: {e.filename}')
+            self._exc = TwisterRunException(f'File not found: {e.filename}')
+        except Exception as e:
+            logger.error('Running simulation failed: %s', e)
+            self._exc = TwisterRunException(e.args)
         else:
-            if self._process.returncode == 0:
-                logger.info('Finished flashing %s', build_dir)
-            elif self._process_ended_with_timeout:
-                raise TwisterTimeoutExpired(f'Process {self._process.pid} terminated after {timeout} seconds')
+            if return_code == 0:
+                logger.info('Running simulation finished with return code %s', return_code)
             else:
-                if stderr := self._process.stderr.read():
-                    logger.error(stderr.decode())
-                raise TwisterFlashException(f'Could not run simulator with PID {self._process.pid}')
-
-    def _collect_process_output(self, process: subprocess.Popen) -> threading.Thread:
-        """Create Thread which saves a process output to a file."""
-        def _read():
-            with process.stdout:
-                for line in iter(process.stdout.readline, b''):
-                    self.queue.put(line.decode().strip())
-
-        return threading.Thread(target=_read, daemon=True)
-
-    def _wait_and_terminate_process(self, process: subprocess.Popen, timeout: float) -> threading.Thread:
-        """Create Thread which kills a process after given time."""
-        def _waiting():
-            end_time = time.time() + timeout
-            while process.poll() is None and time.time() < end_time:
-                time.sleep(0.1)
-            if process.poll() is None:
-                process.kill()
-                logger.error('Process %s terminated after %s seconds', process.pid, timeout)
-                self._process_ended_with_timeout = True
-        return threading.Thread(target=_waiting, daemon=True)
+                logger.warning('Running simulation finished with return code %s', return_code)
+        finally:
+            self.queue.put(END_DATA)  # indicate to the other threads that there will be no more data in queue
+
+    def disconnect(self):
+        pass
+
+    def stop(self) -> None:
+        """Stop device."""
+        self._stop_job = True
+        time.sleep(0.1)  # give a time to end while loop in running simulation
+        if self._process is not None:
```

Comment:
This condition is almost always True. The only situation, when it isn't is when simulation file (e.g. `zephyr.exe`) is not found. As a result `stop` method will try to kill process every time when it was created, even if it was finished correctly. I propose adding here additional verification by:
```python
if self._process is not None and self._process.returncode is not None
```
Maybe thanks to this, try-except below could be removed.

Reply:
It should be rather ` self._process.returncode is None`, because if `returncode` is a integer number it means process already ended. 
I changed this code according to your comment.